rfcs/20240326-block-level-api/README.md

```diff
@@ -0,0 +1,436 @@
+  these instructions (int32 or float).
+- grouped by 64-bit chunks along K (e.g. ARM *MMLA extensions). This
+  stems from the registers actually holding 2d matrices.
+
+Exposing these packed layouts is necessary to benefit from HW
+acceleration. This packing can happen in one of two places:
+- at execution time upon calls to brgemm::execute.
+- ahead of execution time (pre-packing) when one of the matrix is
+  constant.
+
+#### How to express HW specific layouts
+
+There are two main ways oneDNN can expose those layouts:
+- through layout tags. This has the advantage of being very simple,
+  but comes with two limitations:
+    - These tags can contain only information about data layout, not
+      extra metadata (e.g. zero-point related)
+	- it does not scale well when the number of layouts increases.
+- through memory descriptors. This can be more tedious to use as it is
+  an opaque structure that needs to be queried for
+  information. However, it is very flexible as it allows to express
+  non trivial layout descritions, and can be extended with extra
+  information/meta-data.
+
+
+Layout tags here will have to be different than what is used for the
+oneDNN primitive API, as they would mostly map to part of the
+innerblock component of existing memory::desc objects.
+
+The recommendation here would be to go with layout tags only if no
+handling of metadata and extra information is expected from oneDNN
+block level APIs.
+
+#### Query mecanism for layout
+
+Another question is how to inform user about which layout to use for a
+given brgemm problem.  Similar to oneDNN primitive API, we will expose
+methods to query what is the best layout.  Once again, two options are
+possible:
+- the user create a brgemm object, and they query which layout should
+  be used,
+- we expose a static function that the user can use independantly of
+  the brgemm object.
+
+The first option is optimal in the sense that the dispatched
+implementation has all the knowledge necessary to make the layout
+determination. However, it requires the user to pass a brgemm object
+or a layout tag around when the data pre-packing logic is separate to
+the operator implementation.
+
+The second option could be simpler only if the information necessary to
+make layout determination is readily available and does not require to
+pass extra information around (otherwise it would be equivalent to
+pass layout tag / brgemm object around).  Current internal
+implementation of layout selection depends on highest available ISA,
+input datatype, fp_math mode and accumulation datatype.
+
+In general, we would recommend to tie the layout query to a brgemm
+object for two reasons:
+- from oneDNN implementation perspective, we can extend the
+  information used to select best layout
+- in practice, when data is pre-packed, blocking information (so brgemm
+  object creation parameters) have to be known in order to layout data
+  in an optimal way.
```

Comment:
The problem is that the input shape is unknown in the prepacking phase so we are not able to build a right brgemm object then.

Reply:
> The problem is that the input shape is unknown in the prepacking phase so we are not able to build a right brgemm object then.

There are two possibilities here:
- brgemm parameters are independent of your input shapes. For example, you can only use blocks of size 32x32 and 1x32. In that case, you can create the brgemm objects independently of the input shapes, and do blocking with respect to those. This is typically what oneDNN primitives do with [blocked layout ](https://github.com/oneapi-src/oneDNN/blob/e998552539927fccdd21882ba2fbc4788c2991f8/include/oneapi/dnnl/dnnl.hpp#L1177). You can see in those that they are independant of input shape (e.g. AB16b64a uses blocks 16x64 in plain format, AB4b64a4b  uses blocks 16x64 in VNNI layout, ...). 
- brgemm parameters are dependant of your input shapes. In that case, you cannot do efficient pre-packing of the weights. You can only put them in VNNI format or not.

If you are in the latter situation, we can make these static with two implications:
- oneDNN brgemm will never be able to do isa selection based on shapes (e.g. it might be faster to use a lower isa for some shapes that are memory bound).
- you will not get the best performance from pre-packing.

Reply:
@kurapov-peter both options are possible, as laid out above.

Now there is a balance between jitting time (increases with the number of brgemm objects created) and execution time (specialized objects will typically provide better performance). That balance will typically depend on the application.

In any case, hoisting out the kernel creation and memory allocation (e.g. through the use of preallocated memory pools) is desirable to avoid overheads in hot-loops.

Reply:
If `brgemm parameters are independent of your input shapes` can be guaranteed, I vote for option 2 as querying with each brgemm object is redundant. Also, I think this API is to let user understand the preference of brgemm implementation, but not mandatory to follow. User have more understanding about the whole computation looks like and can make the final decision of which layout should be used and pass it during brgemm creation.

Reply:
@mgouicem as we synced offline, we can safely assume that the layout of `B` is independent from the input shapes and we can create brgemm object with arbitrary input shapes, or `M` in particular, to query the layout for `B`, correct? If so, please get these assumptions documented. Thanks.

Reply:
> Also, I think this API is to let user understand the preference of brgemm implementation, but not mandatory to follow. User have more understanding about the whole computation looks like and can make the final decision of which layout should be used and pass it during brgemm creation.

I guess there are 2 parts to the memory layout:
- one aligned with computation blocking. For this one you are right, the brgemm user has better knowledge on which block sizes make sense, and they can use that block size for pre-packing matrices, and for brgemm computation shapes
- one aligned with HW capabilities. This is an implementation details whether the brgemm API  will use instruction requiring plain, packed_32, packed_64 or any other layout.

> as we synced offline, we can safely assume that the layout of B is independent from the input shapes and we can create brgemm object with arbitrary input shapes, or M in particular, to query the layout for B, correct?

Yes we can achieve that. However note that if we do so, we lose opportunity to select isa/layout based on brgemm input shapes. Is that ok ?

Reply:
> Yes we can achieve that. However note that if we do so, we lose opportunity to select isa/layout based on brgemm input shapes. Is that ok ?

If not, how do you suggest we decide the layout in the weight prepacking stage when the input sizes are unknown?

Reply:
> If not, how do you suggest we decide the layout in the weight prepacking stage when the input sizes are unknown?

If shapes are unknown, the layout cannot depend on shape (e.g. same layout is required for gemv and large block gemm). Here it would likely depend on hardware capabilities (e.g. always use `pack_32` when AMX/VNNI is available).